### PR TITLE
FS-4821: loggin configuration changes to give more cloud watch friendly logs

### PR DESCRIPTION
**Pull request by nuwan-samarasinghe**
https://mhclgdigital.atlassian.net/browse/FS-4821

### Change description
In this PR it will format the logging which is going to compatible with cloudwatch logs & this will omit all the sensitive data

- [x] Unit tests and other appropriate tests added or updated
- [x] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### Screenshots of UI changes (if applicable)

image

### CODE REVIEW

**Review by wjrm500 — COMMENTED** [review]
You are a speed demon @nuwan-samarasinghe!

As discussed on call a minute ago, we are now missing some key request information including request method. Also, `request.route.path` is giving the route pattern rather than the route that was actually requested - maybe try `request.path` instead?

One thing I should say is that moving forward we should try to stick within the scope of the Jira ticket, which in this case specified that we want to change the formatting, not the content of the logs. You can request changes to a Jira ticket if you don't think it's right, or request to create a new one if you've identified work that falls outside the scope of another ticket, but ultimately Jira should always remain the source of truth for any development work

File: runner/src/server/plugins/logging.ts [hunk 1]
```diff
@@ -1,4 +1,70 @@
...
+        prettyPrint: false,
+        level: config.logLevel,
+        base: null,
+        formatters: {
+            level: (label) => {
+                return {severity: label.toUpperCase()}; // Align severity levels for CloudWatch
+            },
+        },
+        messageKey: "message",
+        customProps: (request) => ({
```
Comment (wjrm500) [inline runner/src/server/plugins/logging.ts:16]: Does this option actually exist?

File: runner/src/server/plugins/logging.ts [hunk 2]
```diff
@@ -1,4 +1,70 @@
-import pluginLogging from "../../../../digital-form-builder/runner/src/server/plugins/logging";
+import pino from "hapi-pino";
+import {config} from "./utils/AdapterConfigurationSchema";
+
+export default {
+    plugin: pino,
+    options: {
+        prettyPrint: false,
```
Comment (wjrm500) [inline runner/src/server/plugins/logging.ts:7]: If JSON printing is the default maybe it suffices to simply remove any reference to prettyPrint? See [here](https://github.com/pinojs/pino/blob/619e049aed18dd79a9fdfac5dceab86c34e9bda0/docs/pretty.md) where it mentions "_By default, Pino log lines are newline delimited JSON (NDJSON). This is perfect for production usage and long-term storage._"

File: runner/src/server/plugins/logging.ts [hunk 3]
```diff
@@ -1,4 +1,70 @@
...
+                return {severity: label.toUpperCase()}; // Align severity levels for CloudWatch
+            },
+        },
+        messageKey: "message",
+        customProps: (request) => ({
+            requestId: request.info.id, // Include request ID
+            userAgent: request.headers["user-agent"], // Add user agent for context
+            route: request.route.path, // Include route information
+        }),
+        serializers: {
```
Comment (wjrm500) [inline runner/src/server/plugins/logging.ts:21]: If serializers run after hooks, then this makes the inclusion of serializers here a... fallback? Given that req and res are deleted in the hook?

File: runner/src/server/plugins/logging.ts [hunk 4]
```diff
@@ -1,4 +1,70 @@
...
+                            obj.query = obj.req.query;
+                        }
+                        if (obj.req.method) {
+                            obj.method = obj.req.method.toUpperCase();
+                        }
+                        if (obj.req.yar.id) {
+                            obj.requestId = obj.req.yar.id;
+                        }
+                        delete obj.req
+                        delete obj.res
```
Comment (wjrm500) [inline runner/src/server/plugins/logging.ts:50]: This and the line above are redundant given the code just below

File: runner/src/server/plugins/logging.ts [hunk 5]
```diff
@@ -1,4 +1,70 @@
...
+                        if (obj.req.path) {
+                            obj.path = obj.req.path;
+                        }
+                        if (obj.req.query) {
+                            obj.query = obj.req.query;
+                        }
+                        if (obj.req.method) {
+                            obj.method = obj.req.method.toUpperCase();
+                        }
+                        if (obj.req.yar.id) {
```
Comment (wjrm500) [inline runner/src/server/plugins/logging.ts:46]: `obj.req.yar.id` has already been added as `requestId` in the `obj` declaration above

File: runner/src/server/plugins/logging.ts [hunk 6]
```diff
@@ -1,4 +1,70 @@
...
+                        delete obj.res
+                    }
+                    delete obj.req
+                    delete obj.res
+                    return method.apply(this, [obj, ...rest]);
+                }
+                return method.apply(this, inputArgs);
+            },
+        },
+        redact: {
```
Comment (wjrm500) [inline runner/src/server/plugins/logging.ts:59]: Any need for this property now given hook and serializer logic above?

**Review by wjrm500 — COMMENTED** [review]
Hey @nuwan-samarasinghe, thanks for the wait on this. I've left some comments, hope they make some sense

Did you get a chance to check whether the keys logged out here align with the keys logged out in our Python apps?

**Review by wjrm500 — COMMENTED** [review]
Spoke to Nuwan this morn and decided better to overwrite specific properties of the Hapi Pino configuration rather than rewrite completely

**Review by wjrm500 — COMMENTED** [review]
Hey @nuwan-samarasinghe, a few comments...

The scope of this PR has changed over time - as mentioned yesterday we went from a blacklisting approach (currently in `main` branch) via `logRedactPaths`, to a whitelisting approach via a custom `logMethod` hook that only added `requestId`, `method` and `route` to the logs, back to a blacklisting approach (again via `logRedactPaths`) in the current implementation on this PR. At this point I'm happy to stick to the blacklisting approach, but you can appreciate why wholesale changes to functionality across commits in the same PR would be confusing as a reviewer 😺 

Then separately to the redaction of fields, we had some formatting changes including removing pretty printing and making logs "compatible" with CloudWatch, including new values for Hapi Pino logging options `prettyPrint`, `base`, `formatters` and `messageKey`. Now it looks like we've removed all of these completely without explanation, which seems to be a... backwards step? Having pretty printing only turned off via the Copilot manifest means we get different logging behaviour locally, which seems unnecessary and confusing. Further, I'm not sure what the `base` option ever did, but removing the custom `formatters` option means presumably we won't format in the CloudWatch style now, with severity label and capitalisation, and removing `messageKey` also presumably affects adherence to the expected logging style (otherwise why was this specified previously?)

Because of the above uncertainties, I can't approve just yet. A reply to clarify would be appreciated, or we can just have a video call to chat it through

**Comment by nuwan-samarasinghe** [comment]
Hi @wjrm500.....

> **Benefits of the Blacklist and Whitelist Approaches**

>**Blacklist Approach**
> The blacklist approach focuses on excluding specific log messages or sources from being logged. Instead of permitting all logs, it defines a set of loggers, log levels, or patterns that should not appear in the logs. By default, all log messages are allowed unless explicitly excluded.
> 
> **Advantages of the Blacklist Approach:**
> 
> It is particularly useful when most log messages are relevant, and you only need to filter out a small subset of unwanted logs.
> It prevents logging of irrelevant or excessive data, allowing for a more streamlined approach while keeping the configuration manageable.
> 
> **Disadvantages of the Blacklist Approach:**
> 
> It can fall short when certain sources or log types generate excessive noise. In such cases, defining all unwanted patterns can become cumbersome and error-prone.
> **Whitelist Approach**
> The whitelist approach, on the other hand, allows only specific log messages or sources to be logged. Instead of focusing on exclusion, it prioritizes inclusion by specifying exactly which logs are considered important or necessary. Any log that does not match the whitelist criteria is automatically excluded.
> 
> **Advantages of the Whitelist Approach:**
> 
> It helps improve performance and reduces clutter in environments with high logging volume.
> By focusing on meaningful logs, it ensures the log data remains relevant and manageable.


**Why I Chose the Whitelist Approach Initially**

While working on logging configurations for `hapi-pino`, I initially leaned toward the whitelist approach due to challenges I encountered with the default setup. Specifically, I noticed repeated `REDACTED` statements appearing in the logs. These logs lacked value and contributed unnecessary clutter, which undermined the usefulness of the logs.

Since meaningful logs should provide actionable insights, having repetitive and meaningless entries like `REDACTED` added noise. Unfortunately, at that point, I couldn’t find a clear way in the `hapi-pino` documentation to suppress these entries directly. To address this issue, I explored possible solutions and landed on the whitelist approach. By filtering the incoming data and explicitly allowing only relevant log entries, I was able to clean up the logs and ensure their usefulness.

After implementing this solution, I raised a Pull Request (PR) for review. The reviewers provided valuable feedback regarding the methods I used. Based on their suggestions, I revisited my implementation and examined the `hapi-pino` documentation more thoroughly. This time, I aimed to find a way to make minimal changes to the original repository while still resolving the problem.

**Transition Back to the Blacklist Approach**

During my review process, I discovered that we could actually use the blacklist approach same method we were already using in our xGov Forms project. The solution only required a simple change in the configuration properties, which aligned well with our existing setup and prevented significant deviations from the original codebase.

While revisiting the configurations, I also identified another issue: in our AWS environment, we hadn’t disabled the `LOG_PRETTY_PRINT` property. This setting was preventing us from achieving CloudWatch-compatible logging, which is crucial for maintaining proper logging standards in cloud environments.

Ultimately, I switched back to the blacklist approach as it provided a more seamless and less intrusive solution to the problem. The minimal changes ensured we adhered to the original repository’s conventions while addressing the logging issues effectively.

Let me know if you’d like to expand or further refine any section!

**Pull request merged**